### PR TITLE
[CORE] Recursive descent parser + Core/Client separation

### PR DESCRIPTION
**Pull request by AvitalTamir**
This is a complete refactoring of the core Cyphernetes package.
It includes a from-scratch rewrite of the original goyacc Cyphernetes parser and original lexer in favor of a recursive descent function and a more straightforward lexer implementation.
This removes the dependency on goyacc, makes the parser more maintainable and debuggable, and finally lets us produce a coverage report for the parser and lexer.

Additionally, the core package is now separated to packages `core` (which provides the language parser and relationships engine) and `provider` which is an interface for providing client functionality.
The `apiserver` package contains the original Cyphernetes k8s client.
This new architecture will allow external projects to integrate the Cyphernetes language into their project while bringing in their own K8s client or ever an external DB such as Postgres or Elasticsearch (address issue #146)
- [x] pass all existing parser test cases and operator e2e cases.
- [x] lexer test suite rewrite
- [x] include `pkg` in coverage report

### CODE REVIEW

**Review by naorpeled — APPROVED** [review]
LGTM!
great stuff

**Comment by naorpeled** [comment]
btw,
I've noticed that when I use `\n` I'm getting `Error >> error parsing query >> parse error: expected MATCH or CREATE, got {0 \}`, we probably should return a more refined error

**Comment by AvitalTamir** [comment]
Good catch, will address

File: cmd/cyphernetes/main_test.go [hunk 1]
```diff
@@ -102,12 +102,12 @@ func TestCyphernetesShellWithNamespaceFlagHelper(t *testing.T) {
 	fmt.Print(shellPrompt())
 }
 
-func TestCyphernetesShellWithLogLevelFlag(t *testing.T) {
-	if os.Getenv("CI") != "true" {
-		_, stderr := runTestCommand(t, "TestCyphernetesShellWithLogLevelFlagHelper", "TEST_SHELL_LOG_LEVEL")
-		checkOutput(t, stderr, "[DEBUG]", "\"cyphernetes shell -l debug\"")
-	}
-}
+// func TestCyphernetesShellWithLogLevelFlag(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (AvitalTamir) [reply cmd/cyphernetes/main_test.go:105]: Oh yes good catch, not directly related to the PR change. this was a flakey before in CI and failed for me locally when working on the feature. I think a main.go testing issue we already encountered.
Will verify if this is necessary and will either uncomment, remove or reorder tests.

[the remark this answers is not in the document]
Comment (AvitalTamir) [reply cmd/cyphernetes/main_test.go]: Fixed, uncommented back.

**Comment by AvitalTamir** [comment]
@naorpeled I addressed both issues. Test is not uncommented and the error messages improved.
BTW I believe the \n without namespace name was a parsing error in the previous version too, according to the code I fixed :)

**Comment by AvitalTamir** [comment]
BTW your comment also helped me improve the parsing error, changed it from this:
`Error >> error parsing query >> parse error: expected MATCH or CREATE, got {0 \}`

To this:
`Error >> error parsing query >> parse error: expected MATCH or CREATE, got "\"`

**Comment by AvitalTamir** [comment]
@naorpeled This baby is ready 😂 
I think I hashed out every single little issue and this is now fully refactored, complete with engine replacement.
I tested the crap out of it myself, added test cases... Tomorrow I will sit and test this again myself thoroughly.
If you can find the time today/tomorrow - let's get together to go over the PR, maybe you can play with it a little, try to break it and see if there's anything I missed.

**Comment by AvitalTamir** [comment]
To whom it may concern,
This PR is the work of a madman who failed to recognize his own mortality and labored for almost 72h straight over refactoring essentially every single file in an entire monorepo spanning 3 different software packages.
This work has NOT been made possible by AI, which apparently, at the time of writing, is a horrible tool for the job.
I initially thought this would be a 3-4 file job, up until the point I realized the fundamental `FindGVR` function did not belong to the `QueryExecutor` (K8s client) structure I was refactoring. This quickly turned into one of those situations like when you have a tiny bit of loose skin next to your fingernail that you start pulling on and pretty soon you realize you've undone your entire finger, it bleeds pretty bad, you're in a great deal of pain you did NOT see coming and likely need to see a medical professional.

**Comment by elliotxx** [comment]
@AvitalTamir Brother, I noticed that you have accomplished such a feat. It's so cool, it makes it easier for me to integrate [Cyphernetes](https://github.com/AvitalTamir/cyphernetes) in [Karpor](https://github.com/kusionStack/karpor)!

**Comment by elliotxx** [comment]
So far, I am iterating the functionality of v0.6.0 and will need some more time to start integrating Cyphernets...

**Comment by AvitalTamir** [comment]
Hey @elliotxx I was about to reach out to you - I'm about to merge another PR which cleans up this interface further, and includes an integration guide :)
I'll tag you when I do.
If you'd like, we can set up some time to go over this and see if there's anything I can do to help further.

**Comment by AvitalTamir** [comment]
@elliotxx the interface cleanup is now merged - check out the [integration guide](https://github.com/AvitalTamir/cyphernetes/blob/main/docs/INTEGRATION.md) and lmk if there's anything we can do to make this simpler - I'm keen on making this as frictionless as possible.

**Comment by elliotxx** [comment]
@AvitalTamir OK, let me study first. If there is a problem, I will come to you~

**Pull request merged**